### ISSUE TITLE
Flux extraction problem after applying the mask to a dataset

### ISSUE THREAD

**Issue by tstolarczyk** [gammapy/gammapy#3407]
Gammapy 0.17

Hello,
For some "historical reasons" I apply the mask "by hand" on a dataset (essentially because when stacking the mask is not applied on the first dataset).
The original dataset `ds` is build on a `TemplateSpectraModel`.
The excess count list is:

```
ds.excess.data.flatten()
array([11.4, 16.6, 15.2, 63.2, 22.8, 56.4, 24.8, 61. , 57.2, 63.6, 63.6,
       83.4, 63.2, 62.4, 69.6, 49.4, 40. , 51.6, 37.8])

ds.background.data.flatten()
array([ 21.6,  52.4, 129.8, 214.8, 242.2, 192.6, 147.2, 129. , 107.8,
        98.4,  79.4,  49.6,  40.8,  29.6,  20.4,  16.6,  13. ,   6.4,
         5.2])
```
The mask is:
```
ds.mask_safe.data.flatten()
array([False, False, False,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True, False, False,
       False])
```

I use the following "trick" to apply the mask (suggested by @adonath  some time ago):

```
e_true = ds.aeff.energy.edges
e_reco = ds.counts.geom.axes[0].edges
region = ds.counts.geom.region
ds_masked = SpectrumDataset.create(e_true=e_true, e_reco=e_reco, region=region, name=ds.name)
ds_masked.models = ds.models
ds_masked.stack(ds) # Copy the content of ds applyng the mask

```
The excess counts are *almost* the same, the mask is applied as expected:
```
ds_masked.excess.data.flatten()
array([ 0.      ,  0.      ,  0.      , 63.199997, 22.800003, 56.399994,
       24.800003, 61.      , 57.199997, 63.6     , 63.6     , 83.4     ,
       63.2     , 62.4     , 69.6     , 49.4     ,  0.      ,  0.      ,
        0.      ], dtype=float32)
        
ds_masked.background.data.flatten()
array([  0. ,   0. ,   0. , 214.8, 242.2, 192.6, 147.2, 129. , 107.8,
        98.4,  79.4,  49.6,  40.8,  29.6,  20.4,  16.6,   0. ,   0. ,
         0. ], dtype=float32)
```
So far so good. 
Now I extract the flux from `ds `and `ds_masked`. 
When the flux are computed on the `ds_masked`, they are a bit different as expected from the small variation in count number.
But, on the masked dataset the fluxes are not computed on some bins.
The missing bins depend on the seed number (here the seed is 2021, if I use 1965 I have only one flux missing) !
The test notebook is attached.

<table>
<tr>
<th>Original</th>
<th>Masked</th>
</tr>
<tr>
<td>
<pre>
      e_min              e_max                 dnde         
       GeV                GeV            1 / (cm2 GeV s)    
------------------ ------------------ ----------------------
11.000000000000002 13.869761453736817                    nan
13.869761453736817  17.48820752577851                    nan
 17.48820752577851  22.05066060327209                    nan
 22.05066060327209 27.803400224063303  2.854236046457056e-10
27.803400224063303  35.05695715550278  4.818016831854513e-11
 35.05695715550278  44.20287573097234  6.856126758247511e-11
 44.20287573097234  55.73484926888718 2.1002594793257667e-11
 55.73484926888718  70.27536945631296 2.7205093845604733e-11
 70.27536945631296  88.60932822111668 1.2465015023549476e-11
 88.60932822111668 111.72638590934169  9.779805603901128e-12
111.72638590934169 140.87439278642825   6.35111114407704e-12
140.87439278642825   177.626747535253 3.7707196048404215e-12
  177.626747535253 223.96732873791765 2.9538463136314097e-12
223.96732873791765  282.3975839114157 1.3930159851494103e-12
 282.3975839114157  356.0715567239055 1.1944555314991863e-12
 356.0715567239055  448.9661411110258  6.159381673982668e-13
 448.9661411110258  566.0957525467884                    nan
 566.0957525467884  713.7830043452349                    nan
 713.7830043452349  900.0000000000001                    nan
</pre>
</td>
<td>
<pre>
 e_min              e_max                 dnde         
       GeV                GeV            1 / (cm2 GeV s)    
------------------ ------------------ ----------------------
11.000000000000002 13.869761453736817                    nan
13.869761453736817  17.48820752577851                    nan
 17.48820752577851  22.05066060327209                    nan
 22.05066060327209 27.803400224063303  2.854435312149946e-10
27.803400224063303  35.05695715550278                    nan
 35.05695715550278  44.20287573097234                    nan
 44.20287573097234  55.73484926888718 2.1084008448135592e-11
 55.73484926888718  70.27536945631296  2.720242005961813e-11
 70.27536945631296  88.60932822111668 1.2479213743390402e-11
 88.60932822111668 111.72638590934169   9.76892217287604e-12
111.72638590934169 140.87439278642825  6.353487682910813e-12
140.87439278642825   177.626747535253                    nan
  177.626747535253 223.96732873791765  2.956693698008242e-12
223.96732873791765  282.3975839114157  1.393994012170273e-12
 282.3975839114157  356.0715567239055 1.1940941034551917e-12
 356.0715567239055  448.9661411110258  6.161749104158518e-13
 448.9661411110258  566.0957525467884                    nan
 566.0957525467884  713.7830043452349                    nan
 713.7830043452349  900.0000000000001                    nan
</pre>
</td>
</tr>

</table>

What happens?
Thanks!

[FluxProblem-Masking.zip](https://github.com/gammapy/gammapy/files/6753139/FluxProblem-Masking.zip)

**Comment by adonath** [comment]
Thanks @tstolarczyk for reporting this issue. I used the example code that you attached and adapted it to run on the public CTA iris we have in `$GAMMAPY_DATA`. However when doing this I could not reproduce the issue so far...can you maybe check the number of counts stored in the resulting flux points table?

**Comment by tstolarczyk** [comment]
Thanks, @adonath, and sorry for my late reaction I was on vacation.

* The count numbers, as you suggested to print, are the same (exactly in the print, i.e. with one significant digit). They are well the sum of the excess and the background displayed before.
 
* I have also, in the second case, redefine a flux table `fpe2`, instead of using the first one `fep `(juts a try, I did not expect a change...):
```
fpe2 = FluxPointsEstimator(e_edges=axis_reco.edges)
fex2 = fpe2.run(datasets = ds_masked) # Was fep.run
fex2.table["is_ul"] = fex2.table["ts"] < 4
print(fex2.table["e_min","e_max","dnde","counts"])
```
but it does not change anything...

**Comment by tstolarczyk** [comment]
Hello @adonath,
First I have reinstalled gammapy 0.17 on the same anaconda installation and it has not changed anything.

Then I decided to move to gammapy 0.18.2 and everything works perfectly now ! 

In particular, on top of the problem mentioned above, I had erratic fitting problems with Minuit (ExponentialCutoff fit not converging even when outrageously increasing the tolerances on many but not all flux plots). 
All these problems have gone and gammapy seems to behave as expected.

It is a bit frustrating to not have found the problem, but I close this issue.
Thanks
